libp2p/protocols/kademlia/find.nim

```diff
@@ -7,117 +7,141 @@
 # This file may not be copied, modified, or distributed except according to
 # those terms.
 
-import std/[tables, sequtils, sets]
+import std/[tables, sequtils, sets, algorithm]
 import chronos, chronicles, results
-import ../../[peerid, switch, multihash]
+import ../../[peerid, peerinfo, switch, multihash]
 import ../protocol
-import ./[routingtable, lookupstate, protobuf, types]
+import ./[routingtable, protobuf, types]
 
-proc checkConvergence(state: LookupState, me: PeerId): bool {.raises: [], gcsafe.} =
-  let ready = state.activeQueries == 0
-  let noNew = state.selectAlphaPeers().filterIt(me != it).len == 0
-  return ready and noNew
+type LookupState* = object
+  kad: KadDHT
+  targetKey: Key
+  shortlist: Table[PeerId, XorDistance]
+  queried: HashSet[PeerId]
 
-proc waitRepliesOrTimeouts(
-    pendingFutures: Table[PeerId, Future[Message]]
-): Future[(seq[Message], seq[PeerId])] {.async: (raises: [CancelledError]).} =
-  await allFutures(pendingFutures.values.toSeq())
+proc updateShortlist*(
+    state: var LookupState, msg: Message
+): seq[PeerInfo] {.raises: [].} =
+  var newPeerInfos: seq[PeerInfo]
 
-  var receivedReplies: seq[Message] = @[]
-  var failedPeers: seq[PeerId] = @[]
-
-  for (peerId, replyFut) in pendingFutures.pairs:
-    try:
-      receivedReplies.add(await replyFut)
-    except CatchableError:
-      failedPeers.add(peerId)
-      error "Could not send find_node to peer", peerId, err = getCurrentExceptionMsg()
-
-  return (receivedReplies, failedPeers)
-
-proc sendFindNode*(
-    kad: KadDHT, peerId: PeerId, addrs: seq[MultiAddress], targetId: Key
+  for newPeer in msg.closerPeers:
+    let pid = PeerId.init(newPeer.id).valueOr:
+      continue
+    if not state.shortlist.contains(pid):
+      state.shortlist[pid] =
+        xorDistance(pid, state.targetKey, state.kad.rtable.config.hasher)
+
+      let peerInfo = PeerInfo(peerId: pid, addrs: newPeer.addrs)
+      newPeerInfos.add(peerInfo)
+
+  return newPeerInfos
+
+proc sortedShortlist(state: LookupState): seq[(PeerId, XorDistance)] =
+  ## Sort shortlist by closer distance first
+
+  return state.shortlist
+    .pairs()
+    .toSeq()
```

Comment:
would be better if this is one operation (pairs + toSeq) , because memory is trashed unnecessary.

Reply:
would be even better if `shortlist` is heap.